src/Model/ObjectsLoader.php

```diff
@@ -187,10 +187,23 @@ public function hydrateObjects(array $objects): CollectionInterface
      */
     protected function getAssociation(Table $table, string $name): Association|null
     {
-        $associations = $table->associations()->keys();
-        foreach ($associations as $key) {
-            if (strtolower($key) === strtolower($name)) {
-                return $table->getAssociation($key);
+        $associations = $table->associations();
+
+        $assoc = $associations->get($name) ?? $associations->getByProperty($name);
+        if ($assoc !== null) {
+            return $assoc;
+        }
+
+        /*
+         * @todo The following piece of junk should be removed as soon as practically possible.
+         *      It's left here for backwards compatibility with a longstanding bug, but it's wrong.
+         */
+        $name = strtolower($name);
+        foreach ($associations as $assoc) {
+            if (strtolower($assoc->getName()) === $name) {
+                deprecationWarning('Using lowercased association name is a bug and will be removed soon: please use its correct name in CamelCase or the property name in snake_case');
```

Comment:
```suggestion
                deprecationWarning(sprintf('Using the lowercased association name "%s" is a bug and will be removed soon: please use its correct name in CamelCase or the property name in snake_case', $name));
```
Give a hint to which association caused the deprecation notice, even if `$name` has already been lowercased at this point.

Reply:
I see what you want here, but interpolation in a deprecation warning message isn't great in my opinion, as tools aim to aggregate such messages, and CakePHP itself disregards duplicate warning messages.

Interpolation would disrupt both aggregation _and_ deduplication. I'd rather trigger a warning _and_ log a message instead, where the original name could be included to make debugging easier.

Reply:
```suggestion
                deprecationWarning('Using lowercased association name is a bug and will be removed soon: please use its correct name in CamelCase or the property name in snake_case');
                Log::notice(sprintf('Using lowercased association name is a bug and will be removed soon: please use its correct name in CamelCase or the property name in snake_case. The wrong association name is: "%s"', $name));
```
Something like this?

Reply:
Done! Can you please check if the message I've used is clear enough?

Reply:
lgtm